sakura_core/typeprop/CPropTypesKeyHelp.cpp

```diff
@@ -235,7 +235,7 @@ INT_PTR CPropTypesKeyHelp::DispatchEvent(
 					}
 					// 開けたなら1行目を取得してから閉じる -> szAbout
 					std::wstring line=in.ReadLineW();
-					line.copy( szAbout, line.length() + 1, 0 );
+					wcsncpy_s(szAbout, line.c_str(), line.size());
```

Comment:
せっかくなのでサイズを `_TRUNCATE` にしてはどうでしょう？
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strncpy-s-strncpy-s-l-wcsncpy-s-wcsncpy-s-l-mbsncpy-s-mbsncpy-s-l?view=vs-2019

Reply:
そっちのほうが適してそうですね。後で直しておきます。

Reply:
`_TRUNCATE` を設定しない場合であふれた場合にどうなるかですが、[ドキュメント](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strncpy-s-strncpy-s-l-wcsncpy-s-wcsncpy-s-l-mbsncpy-s-mbsncpy-s-l?view=vs-2019)を読むと出力先の先頭にNULL文字が設定される的な事が書かれていたので、まぁ実害は無さそうだけど溢れた場合に空になるよりかは `_TRUNCATE` を設定して途中まででも設定した方が良いかなと思いました。

で、`_TRUNCATE` にする前の `line.size()` の場合であふれた場合にどうなるか一応念の為に動作確認してみました。

Debugビルドで開発環境のデバッガ起動すると下記のダイアログが表示されました。

![image](https://user-images.githubusercontent.com/1131125/80226267-1ccb1680-8687-11ea-8228-2e28f51e885d.png)

`_RETURN_BUFFER_TOO_SMALL` というマクロが、`corecrt_internal_string_templates.h` ファイルの `common_tcsncpy_s` 関数で使われていてそれが assert してるっぽいです。

Releaseビルドで開発環境のデバッガ起動すると `_invoke_watson` で止まりました。

SecureCRT を使う場合は出力先のバッファがコピー元のサイズより小さい場合は、`_TRUNCATE` 指定しないとこういう風になる実装みたいですね。WindowsSDKは 10.0.17763.0 を使ってます

で、[ドキュメント](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strncpy-s-strncpy-s-l-wcsncpy-s-wcsncpy-s-l-mbsncpy-s-mbsncpy-s-l?view=vs-2019#remarks)を読み返してみると、

> otherwise, strDest[0] is set to the null character **and the invalid parameter handler is invoked**, as described in Parameter Validation.

とちゃんと書かれてました。うーん…、お手軽に思えるCRT関数もちゃんと仕様を理解して使わないと駄目ですね。

MSのSecureCRTではなくて一般的なものはどういう仕様なのかを読んでみました。
https://ja.cppreference.com/w/c/string/wide/wcsncpy

切り捨てが発生する場合はエラー扱いで制約ハンドラ関数が呼ばれるんだとか…。
また下記の記述で実行時制約違反にせずに切り捨て動作に出来るようです。
```c
wcsncpy_s(dst, sizeof dst / sizeof *dst, src, (sizeof dst / sizeof *dst)-1);
```

長くなりましたが月並みな結論としてはちゃんと気を付けないといかんですね。:pensive:

Reply:
`_TRUNCATE` が嫌な場合は `line.length()` にしないと落ちますよ、ってことで、
試してみたら落ちましたよ、ってことですね・・・。

Reply:
_TRUNCATE が嫌な、というか良く知らないので最初は使う発想がありませんでした。MS独自なのか分かりませんが訝しんでたというか。

このPRの最初のコミットでは line.length() だったんです。出力先のバッファ溢れケースを動作確認していなくって。

せっかくなので、というアドバイスがberryzplusさんから無かったら欠陥が残るところでした。いやまだあるかもしれませんが。

Reply:
> @berryzplus「_TRUNCATE が嫌な場合は line.length() にしないと落ちますよ、ってことで、試してみたら落ちましたよ、ってことですね・・・。

何から何まで違います。***line.length() にしていたら***切り詰めがエラー扱いで落ちたんです。

_TRUNCATE なら切り詰めてエラーにしない。あなたにとっては怪我の功名のアドバイスだった。

_TRUNCATE が嫌なら ***line.length() にしていたところを書き込み先バッファのサイズ -1 にする***ことで切り詰めつつエラー扱いにもしないで済む。

いいかげんテキストが読めない人間であることを自覚し、動作を確かめなければ愚行を繰り返すことを覚えなさい。これを恥じることがなく行動を改めることもない面の皮にあきれます。

Reply:
> 何から何まで違います。line.length() にしていたら切り詰めがエラー扱いで落ちたんです。

残念です。

目的がずれてるようなのでスルーしときますね :smile:

Reply:
しかし気を付けないと不具合が入ってしまうのは basic_string::copy でも wcsncpy_s でもいっしょですね。固定長バッファが出力先の切り詰めコピー処理は他でもやるだろうし inline 関数を用意した方が良いかも。既にあったりしてw

あと同じようなミスを過去にして兎下さんから指摘されたような気がしなくもない。もう記憶が曖昧で将来同じまたは似たようなミスを起こしそうな気が…。というか他で同じような不具合がないかチェックした方が良いですね。

Reply:
> いいかげんテキストが読めない人間であることを自覚し、動作を確かめなければ愚行を繰り返すことを覚えなさい。これを恥じることがなく行動を改めることもない面の皮にあきれます。

なんども言うようだけど、態度デカ過ぎですよ:smile:

キミの指摘がたいていの場合に意味のあるものであることは事実。
でも、たいていの場合に誰にも理解されずに放置されていることを認識してください。

そして、ほとんどの場合だれかしらを責める文書になっていて、
自分のミスを認めることは滅多にないという点もメンドクサイです。

サクラエディタの開発は、あえてチームで開発しているものです。

本当はたぶん、いったん誰かが引き取ってある程度改造してから共同開発に戻したほうがいいんだけど、あえてチームでやろうとしています。

特定の誰かに敵対するんじゃなく、問題そのものに向き合って対処をしていきたいんです。
これまでgentaやkobakeのせいにして失敗してきたっしょ？
失敗を繰り返すべきじゃないってなら、そういうとこから始めたらいいと思います。

チーム内に勝手に敵を作んな、ってことでよろしく:confused:

Reply:
> しかし気を付けないと不具合が入ってしまうのは basic_string::copy でも wcsncpy_s でもいっしょですね。固定長バッファが出力先の切り詰めコピー処理は他でもやるだろうし inline 関数を用意した方が良いかも。既にあったりしてw

たぶん、あるだろう、ってことで `wstring::copy` の対象は正規表現で洗いました。
おそらく wcsncopy_s もチェックしといたほうがよいと思います。

> あと同じようなミスを過去にして兎下さんから指摘されたような気がしなくもない。もう記憶が曖昧で将来同じまたは似たようなミスを起こしそうな気が…。というか他で同じような不具合がないかチェックした方が良いですね。

結局のばしのばしになっていますが、CDataProfile.h にある StringBufferW を汎化してバッファ書込みの処理を共通化する方向を模索しています。

当面はbv2.4.0の修正リリースが先になるのかな。

Reply:
>> 恥じることがなく行動を改めることもない面の皮にあきれます。
>
> なんども言うようだけど、態度デカ過ぎですよ😄

ここで内省できないところが面の皮の厚さなんだ。

動作確認を怠って何の解決にもならない PR をマージして、そのやり直しであるこの PR でも動作確認を怠って「落ちるよりはマシだと思います。動作確認はしていませんが。」というコメントを残したでしょう。確認したら落ちたんですよ。

この不遜なふるまいをとって態度がでかかったと省みることができないのか。

> そして、ほとんどの場合だれかしらを責める文書になっていて、自分のミスを認めることは滅多にないという点もメンドクサイです。

誰かしらではありません。しかしある一人の視点からはそう見えるかもしれません。そう見える理由があるのでしょう。

> これまでgentaやkobakeのせいにして失敗してきたっしょ？

これはよくわからない。誰が誰のせいにしてきたのか。私は不十分なコード理解に基づく不当であったり一面的であったりする「うんこ」発言に憤ってきました。むしろ環境の変化に起因するバグ報告に対して、そのことで過去の人間に責任を負わせるのはどうかと発言したのが誰だったか。それを読んでいた私にも当の報告者にもそんな認識はなかったにも関わらず、たった一人だけが過去の人間の責任を視野に入れていた。

Reply:
> キミの指摘がたいていの場合に意味のあるものであることは事実。
でも、たいていの場合に誰にも理解されずに放置されていることを認識してください。

そうですね。バグの位置と直し方を指摘したコメントがスルーされました。

Reply:
> 確認したら落ちたんですよ。

それはホントすまんです。

緊急性の高そうな修正だったので外出先から確認もせずにapproveしてました。
緊急性の高い重要な修正さからこそ、自宅に戻って確認をしてからコメントするべきでした。

Reply:
> 誰かしらではありません。しかしある一人の視点からはそう見えるかもしれません。そう見える理由があるのでしょう。

自覚なさそうやなぁと思ってましたが、やっぱりそうですか。

これはちゃいまっせ。

@beru さんとやってるときでも @arigayas さんとやってるときも。
**帰れ、この引きこもり！** と仲裁に入ろうとしたことか:smile:

なんとなく分かると思いますが、ぼくは自分に対する批判には滅法強いタイプです。
なので、ぼくに対してなんか言うのは構わないし、何かの不都合を @berryzplus のせいだ！ってやってくれても構いません。

言葉遣いがぶっきらぼうなのも、相手への配慮ができないのも個性だと思います。
「慣れ合え」とは言いませんが、仲間は大事にしてほしいです。
ここで言いたかったのはそんだけっす。

Reply:
> これはよくわからない。誰が誰のせいにしてきたのか。

2009年以降の SourceForge.net プロジェクトの歴代管理者が、既存コードをできるだけ触りたくないという理由で機能追加の際のプログラム構造の全体的な見直しを怠り、結果として何らかの不具合が発生した責任を取るカタチでプロジェクトから姿を消したことを指して「失敗」と表現しています。

誰が＝2009年以降のプロジェクトの歴代管理者（たくさん居ます）
誰のせい＝既存コード（つまりほぼ genta さんと kobake さん）

途中、別垢で復活して活動してた人もいる気配を感じてますが、「俺のコード」を「俺たちのコード」に昇華させなかったことがプロジェクト衰退の原因だったのではないかと個人的には思っています。「俺のコード」の文化でやってた `SourceForge.net` の息の根を止めたのは、他ならぬぼくなんですけども。

もし機能追加が「俺たちのコード」であったなら、不具合の責任を不当に追及されてプロジェクトを去ることはなかったんじゃないかと思います。

いったい誰が彼らを追い詰めたのかは、ぼくが関わる前の話なので詳しく知りません。